### PR TITLE
feat: add walkDir api

### PR DESCRIPTION
**Pull request by umutuzgur**

### CODE REVIEW

File: lib/fs.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // jshint ignore: start
+/* eslint-disable promise/prefer-await-to-callbacks */
```
Comment (mykola-mokhnach) [inline lib/fs.js]: Is it necessary to disable this check for the whole file?

Comment (umutuzgur) [reply lib/fs.js]: Not really. I can move it to the corresponding lines if you want. There are 2

File: lib/fs.js [hunk 2]
```diff
@@ -55,6 +57,37 @@ let fs = {
       readStream.on('end', () => resolve(fileHash.digest('hex')));
     });
   },
+  /** The callback function which will be called during the directory walking
+   * @name WalkDirCallback
+   * @function
+   * @param {string} itemPath The path of the file
```
Comment (mykola-mokhnach) [inline lib/fs.js]: .. file or folder

File: lib/fs.js [hunk 3]
```diff
@@ -55,6 +57,37 @@ let fs = {
...
+   * @param {boolean} isDirectory Shows if it is a directory or a file
+   * @return {boolean} return true if you wan't to stop wakling
+  */
+
+  /**
+   * Walks a directory given according to the paramters given. The callback will be invoked with a path joined with the dir parameter
+   * @param {string} dir Directory path where we will start walking
+   * @param {boolean} recursive Set it to true if you wan't to continue walking sub directories
+   * @param {WalkDirCallback} callback The callback to be called when a new path is found
```
Comment (mykola-mokhnach) [inline lib/fs.js]: missing description of returned result

Comment (umutuzgur) [reply lib/fs.js]: done

File: lib/fs.js [hunk 4]
```diff
@@ -55,6 +56,38 @@ let fs = {
...
+   * @return {?string} returns the found path or null if the item was not found
+   */
+  async walkDir (dir, recursive, callback) { //eslint-disable-line promise/prefer-await-to-callbacks
+    for (const file of await fs.readdir(dir, { withFileTypes: true })) {
```
Comment (mykola-mokhnach) [inline lib/fs.js]: I would rather rename the var to `item`, since it can be both file or folder

File: lib/fs.js [hunk 5]
```diff
@@ -55,6 +56,38 @@ let fs = {
...
+
+  /**
+   * Walks a directory given according to the paramters given. The callback will be invoked with a path joined with the dir parameter
+   * @param {string} dir Directory path where we will start walking
+   * @param {boolean} recursive Set it to true if you wan't to continue walking sub directories
+   * @param {WalkDirCallback} callback The callback to be called when a new path is found
+   * @return {?string} returns the found path or null if the item was not found
+   */
+  async walkDir (dir, recursive, callback) { //eslint-disable-line promise/prefer-await-to-callbacks
+    for (const item of await fs.readdir(dir, { withFileTypes: true })) {
```
Comment (KazuCocoa) [inline lib/fs.js]: ah, a friend of mine pointed me the `withFileTypes` flag is over 10.10. We support 10+ tho.
Should we update here? cc @umutuzgur

Comment (mykola-mokhnach) [reply lib/fs.js]: @KazuCocoa https://github.com/appium/appium-support/pull/130

Comment (KazuCocoa) [reply lib/fs.js]: ah, okay. thanks!

**Pull request merged**